private/buf/buflsp/mutex.go

```diff
@@ -0,0 +1,171 @@
+	// any value. The address of the global we mint new IDs from is actually great for this,
+	// because users can't access it outside of this package, nor can they extract it out
+	// of the context itself.
+	return context.WithValue(ctx, &nextRequestID, id)
+}
+
+// getRequestID returns the request ID for this context, or 0 if ctx is nil or has no
+// such ID.
+func getRequestID(ctx context.Context) uint64 {
+	if ctx == nil {
+		return 0
+	}
+	id, ok := ctx.Value(&nextRequestID).(uint64)
+	if !ok {
+		return 0
+	}
+
+	// Make sure we don't return 0. This is the only place where the id is actually
+	// witnessed so doing +1 won't affect anything.
+	return id + 1
+}
+
+// mutexPool represents a group of reentrant muteces that cannot be acquired simultaneously.
+//
+// A zero mutexPool is ready to use.
+type mutexPool struct {
+	lock sync.Mutex
+	held map[uint64]*mutex
+}
+
+// NewMutex creates a new mutex in this pool.
+func (mp *mutexPool) NewMutex() mutex {
+	return mutex{pool: mp}
+}
+
+// mutex is a sync.Mutex with some extra features.
+//
+// The main feature is reentrancy-checking. Within the LSP, we need to lock-protect many structures,
+// and it is very easy to deadlock if the same request tries to lock something multiple times.
+// To achieve this, Lock() takes a context, which must be modified by withRequestID().
+type mutex struct {
+	lock sync.Mutex
+	// This is the id of the context currently holding the lock.
+	who  atomic.Uint64
+	pool *mutexPool
+}
+
+// Lock attempts to acquire this mutex or blocks.
+//
+// Unlike [sync.Mutex.Lock], this takes a Context. If that context was updated with withRequestID,
+// this function will panic when attempting to lock the mutex while it is already held by a
+// goroutine using this same context.
+//
+// NOTE: to Lock() and Unlock() with the same context DO NOT synchronize with each other. For example,
+// attempting to lock this mutex from two different goroutines with the same context will
+// result in undefined behavior.
+//
+// Also unlike [sync.Mutex.Lock], it returns an idempotent unlocker function. This can be used like
+// defer mu.Lock()(). Note that only the outer function call is deferred: this is part of the
+// definition of defer. See https://go.dev/play/p/RJNKRcoQRo1. This unlocker can also be used to
+// defer unlocking but also unlock before the function returns.
+//
+// The returned unlocker is not thread-safe.
+func (mu *mutex) Lock(ctx context.Context) (unlocker func()) {
+	var unlocked bool
+	unlocker = func() {
+		if unlocked {
+			return
+		}
+		mu.Unlock(ctx)
+		unlocked = true
+	}
+
+	id := getRequestID(ctx)
+
+	if mu.who.Load() == id && id > 0 {
+		// We seem to have tried to lock this lock twice. Panic, and poison the lock.
+		mu.who.Store(poison)
+		panic("buflsp/mutex.go: non-reentrant lock locked twice by the same request")
+	}
+
+	if mu.pool != nil {
+		mu.pool.lock.Lock()
+		defer mu.pool.lock.Unlock()
```

Comment:
This now deadlocks on any request waiting for a lock as `pool.Lock()` is taken with `lock.Lock()` meaning all locks are blocked by the single lock waiting as the next locker waits on `pool.Lock()` which is blocked by `lock.Lock()`.